x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java

```diff
@@ -0,0 +1,884 @@
+                        + "]"
+                );
+            } else if ((allowedSignatureAlgorithmsHmac.isEmpty() == false) && (isConfiguredJwkSetHmac == false)) {
+                // HMAC signature algorithms present without a HMAC JWKSet
+                throw new SettingsException(
+                    "Settings mismatch. HMAC signature algorithms are set in ["
+                        + allowedSignatureAlgorithmsConfigKey
+                        + "], but no HMAC JWKSet is set in ["
+                        + jwkSetConfigKeyHmac
+                        + "]"
+                );
+            } else if (isConfiguredJwkSetPkc && (allowedSignatureAlgorithmsPkc.isEmpty())) {
+                // RSA/EC JWKSet present without any RSA/EC signature algorithms
+                throw new SettingsException(
+                    "Settings mismatch. RSA/EC JWKSet is set in ["
+                        + jwkSetConfigKeyPkc
+                        + "], but no RSA/EC signature algorithms are set in ["
+                        + allowedSignatureAlgorithmsConfigKey
+                        + "]"
+                );
+            } else if ((allowedSignatureAlgorithmsPkc.isEmpty() == false) && (isConfiguredJwkSetPkc == false)) {
+                // RSA/EC signature algorithms present without an RSA/EC JWKSet
+                throw new SettingsException(
+                    "Settings mismatch. RSA/EC signature algorithms are set in ["
+                        + allowedSignatureAlgorithmsConfigKey
+                        + "], but no RSA/EC JWKSet is set in ["
+                        + jwkSetConfigKeyPkc
+                        + "]"
+                );
+            }
+        }
+
+        // At this point, one or both JWKSets are configured. Load and validate their JWK contents against each set of signature algorithms.
+        JWKSet jwkSetHmac = null;
+        if (isConfiguredJwkSetHmac) {
+            jwkSetHmac = JwtUtil.loadJwkSetFromString(jwkSetConfigKeyHmac, jwkSetContentsHmac);
+            JwtUtil.validateJwkSetContents(jwkSetConfigKeyHmac, jwkSetHmac, allowedSignatureAlgorithmsHmac);
+        }
+        JWKSet jwkSetPkc = null;
+        if (isConfiguredJwkSetPkc) {
+            jwkSetPkc = JwtUtil.loadJwkSetFromString(jwkSetConfigKeyPkc, jwkSetContentsPkc);
+            JwtUtil.validateJwkSetContents(jwkSetConfigKeyPkc, jwkSetPkc, allowedSignatureAlgorithmsPkc);
+        }
+        return new Tuple<>(jwkSetHmac, jwkSetPkc);
+    }
+
+    public static boolean isHttpsScheme(final URI uri) {
+        return (uri != null) && "https".equalsIgnoreCase(uri.getScheme()); // equalsIgnoreCase handles null
+    }
+
+    public static URI parseHttpsUriNoException(final String uriString) {
+        try {
+            final URI uriObj = new URI(uriString);
+            if (JwtUtil.isHttpsScheme(uriObj)) {
+                return uriObj;
+            }
+            LOGGER.trace("Invalid scheme for URI [" + uriString + "]. Only HTTPS is supported.");
```

Comment:
This trace log is required. This method is used to decide if jwkSetPath is a HTTPS URL or file. On Windows, absolute file path triggers an expected exception. On Linux, absolute file path is accepted as a URI, which was unexpected. In other words, I had local file tests passing on Windows but failed in CI on Linux. Root cause is URI parsing accepts relative path, so I had to add an explicit null check. Logs in this method were helpful for me to diagnose the issue, so I would like to leave it in.

Reply:
I understand that you see value in this but I'd have to challenge the fact that since something was helpful to you to debug something during development makes it required to be a log ( even at TRACE level ). 

There are a couple of comments below where I'll try to express my view, but I think a lot of this comes down to personal preference in the end so I won't block/push. I'm going to lay down my thoughts in case it helps

Reply:
Refactored.